### ISSUE TITLE
Provide more information when a file download fails

### ISSUE THREAD

**Issue by alexsuter** [selenide/selenide#2113]
## The problem
In 1 of 10 cases one specific file download failes with:

```
java.io.FileNotFoundException: Failed to download file {by text: test.txt} in 10000 ms.
	at com.codeborne.selenide.impl.Downloads.lambda$firstDownloadedFile$0(Downloads.java:76)
	at java.base/java.util.Optional.orElseThrow(Optional.java:403)
	at com.codeborne.selenide.impl.Downloads.firstDownloadedFile(Downloads.java:75)
	at com.codeborne.selenide.impl.DownloadFileWithProxyServer.clickAndInterceptFileByProxyServer(DownloadFileWithProxyServer.java:84)
	at com.codeborne.selenide.impl.DownloadFileWithProxyServer.lambda$download$0(DownloadFileWithProxyServer.java:46)
	at com.codeborne.selenide.impl.WindowsCloser.runAndCloseArisedWindows(WindowsCloser.java:25)
	at com.codeborne.selenide.impl.DownloadFileWithProxyServer.download(DownloadFileWithProxyServer.java:45)
	at com.codeborne.selenide.commands.DownloadFile.execute(DownloadFile.java:63)
	at com.codeborne.selenide.commands.DownloadFile.execute(DownloadFile.java:29)
	at com.codeborne.selenide.commands.Commands.execute(Commands.java:157)
	at com.codeborne.selenide.impl.SelenideElementProxy.dispatchAndRetry(SelenideElementProxy.java:130)
	at com.codeborne.selenide.impl.SelenideElementProxy.invoke(SelenideElementProxy.java:82)
	at jdk.proxy2/jdk.proxy2.$Proxy29.download(Unknown Source)
	at ch.ivyteam.ivy.project.workflow.webtest.WebTestDownloadCaseDocumentsIT.downloadDocument(WebTestDownloadCaseDocumentsIT.java:70)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:77)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:568)
	at org.junit.platform.commons.util.ReflectionUtils.invokeMethod(ReflectionUtils.java:725)
	at org.junit.jupiter.engine.execution.MethodInvocation.proceed(MethodInvocation.java:60)
	at org.junit.jupiter.engine.execution.InvocationInterceptorChain$ValidatingInvocation.proceed(InvocationInterceptorChain.java:131)
	at org.junit.jupiter.engine.extension.TimeoutExtension.intercept(TimeoutExtension.java:149)
	at org.junit.jupiter.engine.extension.TimeoutExtension.interceptTestableMethod(TimeoutExtension.java:140)
	at org.junit.jupiter.engine.extension.TimeoutExtension.interceptTestMethod(TimeoutExtension.java:84)
	at org.junit.jupiter.engine.execution.ExecutableInvoker$ReflectiveInterceptorCall.lambda$ofVoidMethod$0(ExecutableInvoker.java:115)
	at org.junit.jupiter.engine.execution.ExecutableInvoker.lambda$invoke$0(ExecutableInvoker.java:105)
	at org.junit.jupiter.engine.execution.InvocationInterceptorChain$InterceptedInvocation.proceed(InvocationInterceptorChain.java:106)
	at org.junit.jupiter.engine.execution.InvocationInterceptorChain.proceed(InvocationInterceptorChain.java:64)
	at org.junit.jupiter.engine.execution.InvocationInterceptorChain.chainAndInvoke(InvocationInterceptorChain.java:45)
	at org.junit.jupiter.engine.execution.InvocationInterceptorChain.invoke(InvocationInterceptorChain.java:37)
	at org.junit.jupiter.engine.execution.ExecutableInvoker.invoke(ExecutableInvoker.java:104)
	at org.junit.jupiter.engine.execution.ExecutableInvoker.invoke(ExecutableInvoker.java:98)
	at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.lambda$invokeTestMethod$7(TestMethodTestDescriptor.java:214)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.invokeTestMethod(TestMethodTestDescriptor.java:210)
	at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.execute(TestMethodTestDescriptor.java:135)
	at org.junit.jupiter.engine.descriptor.TestMethodTestDescriptor.execute(TestMethodTestDescriptor.java:66)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$6(NodeTestTask.java:151)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$8(NodeTestTask.java:141)
	at org.junit.platform.engine.support.hierarchical.Node.around(Node.java:137)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$9(NodeTestTask.java:139)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.executeRecursively(NodeTestTask.java:138)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.execute(NodeTestTask.java:95)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1511)
	at org.junit.platform.engine.support.hierarchical.SameThreadHierarchicalTestExecutorService.invokeAll(SameThreadHierarchicalTestExecutorService.java:41)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$6(NodeTestTask.java:155)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$8(NodeTestTask.java:141)
	at org.junit.platform.engine.support.hierarchical.Node.around(Node.java:137)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$9(NodeTestTask.java:139)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.executeRecursively(NodeTestTask.java:138)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.execute(NodeTestTask.java:95)
	at java.base/java.util.ArrayList.forEach(ArrayList.java:1511)
	at org.junit.platform.engine.support.hierarchical.SameThreadHierarchicalTestExecutorService.invokeAll(SameThreadHierarchicalTestExecutorService.java:41)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$6(NodeTestTask.java:155)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$8(NodeTestTask.java:141)
	at org.junit.platform.engine.support.hierarchical.Node.around(Node.java:137)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.lambda$executeRecursively$9(NodeTestTask.java:139)
	at org.junit.platform.engine.support.hierarchical.ThrowableCollector.execute(ThrowableCollector.java:73)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.executeRecursively(NodeTestTask.java:138)
	at org.junit.platform.engine.support.hierarchical.NodeTestTask.execute(NodeTestTask.java:95)
	at org.junit.platform.engine.support.hierarchical.SameThreadHierarchicalTestExecutorService.submit(SameThreadHierarchicalTestExecutorService.java:35)
	at org.junit.platform.engine.support.hierarchical.HierarchicalTestExecutor.execute(HierarchicalTestExecutor.java:57)
	at org.junit.platform.engine.support.hierarchical.HierarchicalTestEngine.execute(HierarchicalTestEngine.java:54)
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.execute(EngineExecutionOrchestrator.java:107)
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.execute(EngineExecutionOrchestrator.java:88)
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.lambda$execute$0(EngineExecutionOrchestrator.java:54)
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.withInterceptedStreams(EngineExecutionOrchestrator.java:67)
	at org.junit.platform.launcher.core.EngineExecutionOrchestrator.execute(EngineExecutionOrchestrator.java:52)
	at org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:114)
	at org.junit.platform.launcher.core.DefaultLauncher.execute(DefaultLauncher.java:86)
	at org.junit.platform.launcher.core.DefaultLauncherSession$DelegatingLauncher.execute(DefaultLauncherSession.java:86)
	at org.apache.maven.surefire.junitplatform.LazyLauncher.execute(LazyLauncher.java:55)
	at org.apache.maven.surefire.junitplatform.JUnitPlatformProvider.execute(JUnitPlatformProvider.java:223)
	at org.apache.maven.surefire.junitplatform.JUnitPlatformProvider.invokeAllTests(JUnitPlatformProvider.java:175)
	at org.apache.maven.surefire.junitplatform.JUnitPlatformProvider.invoke(JUnitPlatformProvider.java:139)
	at org.apache.maven.surefire.booter.ForkedBooter.runSuitesInProcess(ForkedBooter.java:456)
	at org.apache.maven.surefire.booter.ForkedBooter.execute(ForkedBooter.java:169)
	at org.apache.maven.surefire.booter.ForkedBooter.run(ForkedBooter.java:595)
	at org.apache.maven.surefire.booter.ForkedBooter.main(ForkedBooter.java:581)
```

## Details
The API doc outlines when a FileNotFoundException is thrown, that there was a 40X error. But I can not see any 40X request in my generated HAR report.

I tried a lot of things, but I really don't get it. Can you help me to get more insights? Or can you provide more details, why the file download has failed?

## Tell us about your environment
* **Selenide Version**: 
6.9.0

## Code To Reproduce Issue 

```java
openView("cases.xhtml");
$(byText("Created case of TestData")).shouldBe(visible).click();
$("#caseName").shouldBe(visible).shouldHave(text("Created case of TestData"));
$(".documents-card").shouldHave(text("test.txt"));
var downloadDocumentElement = $(byText("test.txt")).shouldBe(visible);
Selenide.screenshot("download0");
var scrolledDocumentElement = downloadDocumentElement.scrollIntoView("{behavior:\"smooth\",block:\"center\"}");
$(".documents-card").scrollTo();
Selenide.screenshot("download1");

// This failes
File download = scrolledDocumentElement.download();

assertThat(download).hasName("test.txt");
assertThat(Files.readString(download.toPath())).isEqualTo("this is test document");
```

**Comment by asolntsev** [comment]
@alexsuter We would like to add more information to the error message, but the problem is that often there is no more information. :( 

But first of all, did you read Selenide logs? did you see some log messages before the failure?

**Comment by alexsuter** [comment]
> @alexsuter We would like to add more information to the error message, but the problem is that often there is no more information. :(
> 
> But first of all, did you read Selenide logs? did you see some log messages before the failure?
> 
> For example, I see logs like this in my console. They contain lots of useful information and might give hints about the downloading process.
> 
> ```
> 22:22:57:700 [main] INFO WebDriverThreadLocalContainer - No webdriver is bound to current thread: 1 - let's create a new webdriver
> 22:22:57:709 [main] DEBUG CreateDriverCommand - Creating webdriver in thread 1 (ip: 192.168.0.21, host: Andreis-MacBook-Pro.local)...
> 
> 22:22:58:166 [main] INFO DefaultHttpProxyServer - Starting proxy at address: 0.0.0.0/0.0.0.0:0
> 22:22:58:175 [main] INFO DefaultHttpProxyServer - Proxy listening with TCP transport
> 22:22:58:202 [main] INFO DefaultHttpProxyServer - Proxy started at address: /[0:0:0:0:0:0:0:0]:56685
> 22:22:58:207 [main] INFO SelenideProxyServer - Using proxy host resolved by Selenium: '192.168.0.21' (fyi BrowserUpProxy resolved : '127.0.0.1')
> 22:22:58:209 [main] INFO FileHelper - Creating folder: /Users/andrei/projects/selenide/modules/proxy/build/downloads/1673036578207_81834_1
> 
> downloadsFolder=/Users/andrei/projects/selenide/modules/proxy/build/downloads/1673036578207_81834_1
> 22:22:58:283 [main] INFO WebDriverManager - Using chromedriver 108.0.5359.71 (resolved driver for Chrome 108)
> 22:22:58:294 [main] INFO WebDriverManager - Exporting webdriver.chrome.driver as /Users/andrei/.cache/selenium/chromedriver/mac-arm64/108.0.5359.71/chromedriver
> 
> 22:22:58:863 [main] INFO WebDriverFactory - BrowserName=chrome Version=108.0.5359.124 Platform=MAC
> 22:22:58:863 [main] INFO WebDriverFactory - Selenium WebDriver v. 4.7.2 build revision: 4d4020c3b7
> 22:22:58:864 [main] INFO CreateDriverCommand - Created webdriver in thread 1: ChromeDriver -> ChromeDriver: chrome on MAC (4047220fa8f6ff962669bb40ca99f1a0)
> 22:22:58:864 [main] INFO DisposablesRegistry - Register WebDriverInstance[config=StaticConfig,webDriver=ChromeDriver: chrome on MAC (4047220fa8f6ff962669bb40ca99f1a0),proxy=Selenide proxy server: 56685,downloadsFolder=/Users/andrei/projects/selenide/modules/proxy/build/downloads/1673036578207_81834_1] in 1 [size=1]
> 
> 22:22:59:429 [main] DEBUG Plugins - Using default implementation of com.codeborne.selenide.impl.DownloadFileToFolder: com.codeborne.selenide.impl.DownloadFileToFolder
> 
> 22:22:59:454 [main] DEBUG DownloadFile - Download file: method: PROXY, timeout: 1000 ms, filter: with extension "rar"
> 22:22:59:598 [qtp391877669-23] INFO BaseHandler - 10:01:59:598 GET https://127.0.0.1:25066/files/hello_world.txt -> 200 1 ms
> 22:22:59:599 [LittleProxy-0-ProxyToServerWorker-0] DEBUG FileDownloadFilter - Intercepted response #1: https://127.0.0.1:25066/files/hello_world.txt -> 200 "OK" {date=Fri, 06 Jan 2023 20:22:59 GMT, content-length=14, server=Jetty(9.4.50.v20221201), content-disposition=attachment; filename=hello_world.txt, content-type=text/html, cache-control=no-cache} text/html  (14 bytes)
> 
> 22:22:59:599 [LittleProxy-0-ProxyToServerWorker-0] INFO FileHelper - Creating folder: /Users/andrei/projects/selenide/modules/proxy/build/downloads/c8011f70-cfc9-49d8-91f4-5dca8408e3da
> 
> 22:23:00:682 [main] INFO DownloadFileWithProxyServer - Downloaded 1 files:
>   #1  /Users/andrei/projects/selenide/modules/proxy/build/downloads/c8011f70-cfc9-49d8-91f4-5dca8408e3da/hello_world.txt
> 
> 22:23:00:683 [main] INFO DownloadFileWithProxyServer - Just in case, intercepted 1 responses:
>   #1  https://127.0.0.1:25066/files/hello_world.txt -> 200 "OK" {date=Fri, 06 Jan 2023 20:22:59 GMT, content-length=14, server=Jetty(9.4.50.v20221201), content-disposition=attachment; filename=hello_world.txt, content-type=text/html, cache-control=no-cache} text/html  (14 bytes)
> 
> 22:23:00:720 [main] INFO FileHelper - Creating folder: /Users/andrei/projects/selenide/modules/proxy/build/reports/tests/integration/FileDownloadViaProxyTest/downloadsFiles
> 
> java.io.FileNotFoundException: Failed to download file with extension "rar" in 1000 ms.
> 	at com.codeborne.selenide.impl.Downloads.lambda$firstDownloadedFile$0(Downloads.java:76)
> ```

Thanks! I'm struggling to get this logs in my build.

**Comment by asolntsev** [comment]
@alexsuter This might help: https://github.com/selenide/selenide/wiki/slf4j
